File: archive/cdktf-py/Pr4943/lib/networking_stack.py
"""Networking infrastructure stack."""

from constructs import Construct
from cdktf_cdktf_provider_aws.vpc import Vpc
from cdktf_cdktf_provider_aws.subnet import Subnet
from cdktf_cdktf_provider_aws.internet_gateway import InternetGateway
from cdktf_cdktf_provider_aws.nat_gateway import NatGateway
from cdktf_cdktf_provider_aws.eip import Eip
from cdktf_cdktf_provider_aws.route_table import RouteTable
from cdktf_cdktf_provider_aws.route import Route
from cdktf_cdktf_provider_aws.route_table_association import RouteTableAssociation


class NetworkingStack(Construct):
    """Networking infrastructure for healthcare platform."""

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        environment_suffix: str
    ):
        """Initialize networking stack."""
        super().__init__(scope, construct_id)

        # Create VPC
        self.vpc = Vpc(
            self,
            "vpc",
            cidr_block="10.0.0.0/16",
            enable_dns_hostnames=True,
            enable_dns_support=True,
            tags={
                "Name": f"healthcare-vpc-{environment_suffix}"
            }
        )

        # Create public subnets in multiple AZs
        self.public_subnet_1 = Subnet(
            self,
            "public_subnet_1",
            vpc_id=self.vpc.id,
            cidr_block="10.0.1.0/24",
            availability_zone="us-east-1a",
            map_public_ip_on_launch=True,
            tags={
                "Name": f"healthcare-public-subnet-1-{environment_suffix}"
            }
        )

        self.public_subnet_2 = Subnet(
            self,
            "public_subnet_2",
            vpc_id=self.vpc.id,
            cidr_block="10.0.2.0/24",
            availability_zone="us-east-1b",
            map_public_ip_on_launch=True,
            tags={
                "Name": f"healthcare-public-subnet-2-{environment_suffix}"
            }
        )

        # Create private subnets in multiple AZs
        self.private_subnet_1 = Subnet(
            self,
            "private_subnet_1",
            vpc_id=self.vpc.id,
            cidr_block="10.0.10.0/24",
            availability_zone="us-east-1a",
            tags={
                "Name": f"healthcare-private-subnet-1-{environment_suffix}"
            }
        )

        self.private_subnet_2 = Subnet(
            self,
            "private_subnet_2",
            vpc_id=self.vpc.id,
            cidr_block="10.0.11.0/24",
            availability_zone="us-east-1b",
            tags={
                "Name": f"healthcare-private-subnet-2-{environment_suffix}"
            }
        )

        # Create Internet Gateway
        self.igw = InternetGateway(
            self,
            "igw",
            vpc_id=self.vpc.id,
            tags={
                "Name": f"healthcare-igw-{environment_suffix}"
            }
        )

        # Create Elastic IP for NAT Gateway
        self.nat_eip = Eip(
            self,
            "nat_eip",
            domain="vpc",
            tags={
                "Name": f"healthcare-nat-eip-{environment_suffix}"
            }
        )

        # Create NAT Gateway in public subnet
        self.nat_gateway = NatGateway(
            self,
            "nat_gateway",
            allocation_id=self.nat_eip.id,
            subnet_id=self.public_subnet_1.id,
            tags={
                "Name": f"healthcare-nat-gateway-{environment_suffix}"
            }
        )

        # Create route table for public subnets
        self.public_route_table = RouteTable(
            self,
            "public_route_table",
            vpc_id=self.vpc.id,
            tags={
                "Name": f"healthcare-public-rt-{environment_suffix}"
            }
        )

        # Create public route to internet gateway
        Route(
            self,
            "public_route",
            route_table_id=self.public_route_table.id,
            destination_cidr_block="0.0.0.0/0",
            gateway_id=self.igw.id
        )

        # Associate public subnets with public route table
        RouteTableAssociation(
            self,
            "public_subnet_1_association",
            subnet_id=self.public_subnet_1.id,
            route_table_id=self.public_route_table.id
        )

        RouteTableAssociation(
            self,
            "public_subnet_2_association",
            subnet_id=self.public_subnet_2.id,
            route_table_id=self.public_route_table.id
        )

        # Create route table for private subnets
        self.private_route_table = RouteTable(
            self,
            "private_route_table",
            vpc_id=self.vpc.id,
            tags={
                "Name": f"healthcare-private-rt-{environment_suffix}"
            }
        )

        # Create private route to NAT gateway
        Route(
            self,
            "private_route",
            route_table_id=self.private_route_table.id,
            destination_cidr_block="0.0.0.0/0",
            nat_gateway_id=self.nat_gateway.id
        )

        # Associate private subnets with private route table
        RouteTableAssociation(
            self,
            "private_subnet_1_association",
            subnet_id=self.private_subnet_1.id,
            route_table_id=self.private_route_table.id
        )

        RouteTableAssociation(
            self,
            "private_subnet_2_association",
            subnet_id=self.private_subnet_2.id,
            route_table_id=self.private_route_table.id
        )

    @property
    def vpc_id(self):
        """Return VPC ID."""
        return self.vpc.id

    @property
    def public_subnet_ids(self):
        """Return list of public subnet IDs."""
        return [self.public_subnet_1.id, self.public_subnet_2.id]

    @property
    def private_subnet_ids(self):
        """Return list of private subnet IDs."""
        return [self.private_subnet_1.id, self.private_subnet_2.id]

File: archive/pulumi-ts/Pr6004/lib/networking-stack.ts
/**
 * networking-stack.ts
 *
 * VPC and networking infrastructure across 3 availability zones.
 */
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface NetworkingStackArgs {
  environmentSuffix: string;
  tags: pulumi.Input<{ [key: string]: string }>;
}

export class NetworkingStack extends pulumi.ComponentResource {
  public readonly vpc: aws.ec2.Vpc;
  public readonly publicSubnetIds: pulumi.Output<string>[];
  public readonly privateSubnetIds: pulumi.Output<string>[];

  constructor(
    name: string,
    args: NetworkingStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:networking:NetworkingStack', name, args, opts);

    const { environmentSuffix, tags } = args;

    // Get availability zones
    const availabilityZones = aws.getAvailabilityZonesOutput({
      state: 'available',
      filters: [
        {
          name: 'region-name',
          values: ['us-east-1'],
        },
      ],
    });

    // VPC
    this.vpc = new aws.ec2.Vpc(
      `vpc-${environmentSuffix}`,
      {
        cidrBlock: '10.0.0.0/16',
        enableDnsHostnames: true,
        enableDnsSupport: true,
        tags: {
          Name: `vpc-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Internet Gateway
    const igw = new aws.ec2.InternetGateway(
      `igw-${environmentSuffix}`,
      {
        vpcId: this.vpc.id,
        tags: {
          Name: `igw-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Public Subnets (3 AZs)
    const publicSubnets: aws.ec2.Subnet[] = [];
    const azNames = availabilityZones.names;

    for (let i = 0; i < 3; i++) {
      const subnet = new aws.ec2.Subnet(
        `public-subnet-${i + 1}-${environmentSuffix}`,
        {
          vpcId: this.vpc.id,
          cidrBlock: `10.0.${i}.0/24`,
          availabilityZone: azNames[i],
          mapPublicIpOnLaunch: true,
          tags: {
            Name: `public-subnet-${i + 1}-${environmentSuffix}`,
            ...tags,
          },
        },
        { parent: this }
      );
      publicSubnets.push(subnet);
    }

    // Private Subnets (3 AZs)
    const privateSubnets: aws.ec2.Subnet[] = [];
    for (let i = 0; i < 3; i++) {
      const subnet = new aws.ec2.Subnet(
        `private-subnet-${i + 1}-${environmentSuffix}`,
        {
          vpcId: this.vpc.id,
          cidrBlock: `10.0.${i + 10}.0/24`,
          availabilityZone: azNames[i],
          mapPublicIpOnLaunch: false,
          tags: {
            Name: `private-subnet-${i + 1}-${environmentSuffix}`,
            ...tags,
          },
        },
        { parent: this }
      );
      privateSubnets.push(subnet);
    }

    // NAT Gateway (single for cost optimization)
    const eip = new aws.ec2.Eip(
      `nat-eip-${environmentSuffix}`,
      {
        domain: 'vpc',
        tags: {
          Name: `nat-eip-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    const natGateway = new aws.ec2.NatGateway(
      `nat-gateway-${environmentSuffix}`,
      {
        allocationId: eip.id,
        subnetId: publicSubnets[0].id,
        tags: {
          Name: `nat-gateway-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    // Public Route Table
    const publicRouteTable = new aws.ec2.RouteTable(
      `public-rt-${environmentSuffix}`,
      {
        vpcId: this.vpc.id,
        tags: {
          Name: `public-rt-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    new aws.ec2.Route(
      `public-route-${environmentSuffix}`,
      {
        routeTableId: publicRouteTable.id,
        destinationCidrBlock: '0.0.0.0/0',
        gatewayId: igw.id,
      },
      { parent: this }
    );

    // Associate public subnets with public route table
    publicSubnets.forEach((subnet, i) => {
      new aws.ec2.RouteTableAssociation(
        `public-rta-${i + 1}-${environmentSuffix}`,
        {
          subnetId: subnet.id,
          routeTableId: publicRouteTable.id,
        },
        { parent: this }
      );
    });

    // Private Route Table
    const privateRouteTable = new aws.ec2.RouteTable(
      `private-rt-${environmentSuffix}`,
      {
        vpcId: this.vpc.id,
        tags: {
          Name: `private-rt-${environmentSuffix}`,
          ...tags,
        },
      },
      { parent: this }
    );

    new aws.ec2.Route(
      `private-route-${environmentSuffix}`,
      {
        routeTableId: privateRouteTable.id,
        destinationCidrBlock: '0.0.0.0/0',
        natGatewayId: natGateway.id,
      },
      { parent: this }
    );

    // Associate private subnets with private route table
    privateSubnets.forEach((subnet, i) => {
      new aws.ec2.RouteTableAssociation(
        `private-rta-${i + 1}-${environmentSuffix}`,
        {
          subnetId: subnet.id,
          routeTableId: privateRouteTable.id,
        },
        { parent: this }
      );
    });

    // Export subnet IDs
    this.publicSubnetIds = publicSubnets.map(s => s.id);
    this.privateSubnetIds = privateSubnets.map(s => s.id);

    this.registerOutputs({
      vpcId: this.vpc.id,
      publicSubnetIds: this.publicSubnetIds,
      privateSubnetIds: this.privateSubnetIds,
    });
  }
}
